---
layout: default
title: "Walkthrough"
date: "2025-01-11"
permalink: /walkthrough/

---
# In Class Walkthrough 

![image](https://github.com/user-attachments/assets/26279a58-b4ba-4dfc-8691-84ff4936b375)


- The "hands-on" portion of the class will focus on trying to manipulate 16S abundance tables so that we can pull out some cool data and figures!

- We won't have time to go through the entirety of the 16S analysis pipeline. I've already pre-processed the raw 16S reads to do the following: 1) Import raw 16S reads into Qiime2, 2) Trim primers and adapters, 3) denoising data (forming ASVs), 4) generating a feature abundance table, 5) assigning ASV taxonomy

- If you are interested in how to do this, the code is available in the "data pre-processing" section

- The rest of the analyses that we will be going through together will be in R. Don't worry if you don't have much R experience. Most of the code below can be directly copied and pasted. If that is still an issue and/or package installation proves to be challenging, you can pair up with someone else in the class. The main goal here is not necessarily to learn the exact R syntax, but rather to understand the analysis workflow and to get a grasp of the types of information we can and cannot get from 16S sequencing data

- Record your answers as you go in a word doc and then submit them here when you're ready! [Submit here:](https://docs.google.com/forms/d/e/1FAIpQLSdz0U93_kgIs8GinZKoUH6nPwwIGTbZFKZtH6h0Cu_VsrW0Zg/viewform?usp=dialog)


## Importing the data 

- We will exclusively be working in R, but before we do so it's important to understand the data types we will be working with and importing

- I used Qiime2 (see resources) for pre-processing, which involves storing data in file types called "artifacts" in .qza files. We can import the artifacts into R and convert them into a phyloseq object (this is will make more sense below)

- Phyloseq is an R package that integrates: 1) abundance tables (ie- the abundance of each ASV in each sample), 2) taxonomy information (ie - what is the taxonomic identity of each ASV?), 3) metadata (what are features of the samples themselves?), 4) phylogeny information (what is the phylogenetic relatedness of the ASVs in the dataset?)

- These 4 features can be integrated into a single phyloseq "object" in R. Having all this information integrated into a single object makes our code cleaner and more straightforward to run :) 


- Please download the following artifacts and files into the same directory: 1)  [Feature abundance artifact](https://github.com/anshulsinha1/617_16S_Tutorial/blob/main/Resources/feature-sample-filtered-table.qza), 2) [Taxonomy artifact](https://github.com/anshulsinha1/617_16S_Tutorial/blob/main/Resources/filtered-taxonomy.qza), 3) [metadata](https://github.com/anshulsinha1/617_16S_Tutorial/blob/main/Resources/sample-metadata.txt) 4) [phylogeny artifact](https://github.com/anshulsinha1/617_16S_Tutorial/blob/main/Resources/rooted-tree.qza)


### Installing R packages 

- Our first step will be integrating these artifacts into a phyloseq object

- Before we do that, we'll have to install some R packages. These will include some basic R libraries for data manipulation and visualization (dplyr and ggplot2), phyloseq, and packages that use phyloseq (qiime2R, microViz, ape, microbiome)

- Below are instructions to install these libraries. If you can, try to install these before class. We'll still spend a bit of time ensuring we are all setup properly

```{r}


###Install phyloseq###

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("phyloseq")


###Install ape###

install.packages("ape")


###Install qiime2R###

if (!requireNamespace("devtools", quietly = TRUE)){install.packages("devtools")}
devtools::install_github("jbisanz/qiime2R")


###Install Microviz###

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(c("phyloseq", "microbiome", "ComplexHeatmap"), update = FALSE)

install.packages(
  "microViz",
  repos = c(davidbarnett = "https://david-barnett.r-universe.dev", getOption("repos"))
)


###Install Microbiome###

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("microbiome")


###Install vegan###
install.packages("vegan")


```



- Now load the installed packages!

```{r}
library(phyloseq)
library(qiime2R)
library(microViz)
library(ggplot2)
library(ape)
library(microbiome)
library(dplyr)
library(vegan)
```

### Importing artifacts into a phyloseq object

- Now that our packages are installed and loaded, we can import our Qiime2 artifacts and create a phyloseq object in one simple step using qza_to_phyloseq
- Simply replace the file-path in quotations with the file-path in your directory (right-click, option, copy path-name on mac)


```{r}
physeq <- qza_to_phyloseq(
    features="/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/feature-sample-filtered-table.qza",
    tree="/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/rooted-tree.qza",
    taxonomy="/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/filtered-taxonomy.qza",
    metadata = "/Users/anshulsinha/Desktop/microbiome_617_lectures/16S_workshop/artifacts/qza_to_phyloseq/sample-metadata.txt"
    )
```


## Section 1: Exploring the Data

-  We've now succesfully imported our data into a phyloseq object!
-  View the phyloseq object to get an idea of the data we are working with

```{r}
physeq
```


-  **Question 1A) How many taxa (ASVs) and samples do we have in the dataset?**


- Now lets explore the metadata a bit to get a sense of the dataset-specific parameters we're dealing with
- We can access different features of the phyloseq object
- To acess the metadata specifically we can do the following:

```{r}
sample_data(physeq)

#and save it as a dataframe 
metadata <- as.data.frame(sample_data(physeq))
```

```{r}
#You can use unique() in base R to find all unique values in a specific column 
#For example
unique(metadata$Day)

```

- **Question 1B) How many time points are in the experiment?**
- **Question 1C) What are the three diet treatments the mice are given?**
- **Question 1D) Where are the mouse vendors from?**
 
Start to think of some microbiome-based questions one could answer based on the metdata we have!


## Section 2: Exploring Microbiome Composition

- Now that we have a sense of the samples we are working with, lets try to visualize some data
- A good way to get a first overview of microbiome data is through with simple bar plots. These aren't perfect, but they can give us a nice first impression of what to expect
- We can make a simple bar plot using the Microviz comp_barplot command 
- This command will plot (per sample) the relative abundance of bacteria at different taxonomic levels 
- Lets first look at some high-level taxonomy. 
- First we need to "fix" the names taxa - removes possible unclassified ("NAs") from our dataset, which are incompatible with the comp_barplot function

```{r}
fixed_physeq <- tax_fix(physeq)
fixed_physeq <- fixed_physeq %>% 
  tax_fix(unknowns = c("uncultured"))
```

- Our "fixed" phyoloseq can now be used to generate relative abundance plots downstream 
- Now plot using the comp_barplot function. Use the "fixed" phyloseq object and toggle between different taxonomic ranks using the "tax_level" flag 
- Spend some time looking through these plots at different taxonomic levels 

```{r}
comp_barplot(fixed_physeq, tax_level = "Phylum") 
```

![image](https://github.com/user-attachments/assets/b7a2bbe8-2535-46f0-9305-727b4be019c9)


-  Note that by default, comp_barplot will plot the most 8 abundant taxa. You can increase this number by using the n_taxa flag. This is more useful for lower level ranks
-  
```{r}
comp_barplot(fixed_physeq, tax_level = "Phylum", n_taxa = 12) 
```


Does anything stand out here? 
- **Question 2A) Which bacterial phyla tend to be the most abundant in the dataset?**
- **Question 2B) What about at the family and genus levels?**
- **Question 2C) Try doing this at species level, anything strange to report of the species names?**



## Part 3 - Exploring ASV trends 

- We now have a nice overview of the high level taxonomy in the data, but what about at a finer resolution? 
- Since we have a large number of distinct ASVs in our dataset, it's hard to visualize the data using something like a bar plot or even a heatmap
- We can see how prevalent each of these ASVs are across the dataset to get a sense of which ones are "important"
- This will require a bit of data manipulation... The code below transforms our phyloseq object to a dataframe, which gives the number of samples that each ASV is found in

  
```{r}
#For folks with R experience: we are applying a function that 1) accesses the OTU table from the Phyloseq, 2) MARGIN =1 defines that our taxa are in the rows, 3) FUN: for each row (ASV) it counts the number of non-zero entries 


prevalence = apply(X = otu_table(physeq),MARGIN = 1, FUN = function(x){sum(x > 0)})


#convert to a df
prevalence.df <- as.data.frame(prevalence)

# Our ASVs (rows) have indecipherable codes (used by the Silva database)!
# Lets use our taxonomy table to help us find out the ID of these 




# 1) Get the taxonomy table from the phyloseq object! 
tax_df <- as.data.frame(tax_table(physeq))


#2) now merge the taxonomy dataframe and the prevalence dataframe 
merged_df_prevalence <- merge(prevalence.df, tax_df, by = "row.names", all = TRUE)


```

- Now we have a dataframe that has each unique ASV as a row name, taxonomy labels, and how many samples each ASV is found in.

- Before we think about ASV prevalence, take a look at the dataframe and look at the taxonomy annotations. Notice that we have NO species-level assignments. This is not a mistake. In general, V3-V4 primers don't provide us with good taxonomic resolution at the species level. The Silva database also doesn't curate species names in their database, so for accuracy sake we won't consider these. 


- You can use sum(!is.na(df$your_column)) to count the number of rows that are NOT NA (ie- that have a taxonomy assignment). Ie - 0 for species



```{r}
sum(!is.na(merged_df_prevalence$Species))
```

**Question 3A) How many ASVs have genera and family-level designations?**



- Now we can dive into some of the prevalence data a bit more
- Order the order the dataframe in descending order to get an idea of the most prevalent ASVs


```{r}
df_prevalence_ordered <- merged_df_prevalence[order(-merged_df_prevalence$prevalence), ]
```

**Question 3B) How many ASVs are found in ALL 152 samples and what genera do they belong to?**


- Let's do something similar, but with abundance instead of prevalence!

```{r}
# Step 1: Transform the phyloseq object to get relative abundance info for each ASV using the "transform_sample_counts" function
#Now we have a phyloseq object that has relative abundance counts instead of "raw counts"

physeq_rel <- transform_sample_counts(physeq, function(x) x / sum(x))



# Step 2: Extract abundance data from our phyloseq object 

taxa_abundance <- taxa_sums(physeq_rel)  # Sum the relative abundance for each ASV

#Turn into a dataframe
abundance_df <- data.frame(ASV = names(taxa_abundance), Abundance = taxa_abundance)



# Step 3: Merge with our taxonomy dataframe
merged_df_abundance <- merge(abundance_df, tax_df, by = "row.names", all = TRUE)

#Now we have 


# Step 4: Rank taxa by abundance
#ordering the dataframe by abundance 

merged_df_abundance <- merged_df_abundance[order(-merged_df_abundance$Abundance), ]
#creating a rank using seq_len, which gives a integer in order to the ordered row names
merged_df_abundance$Rank <- seq_len(nrow(merged_df_abundance))


```

- Now we can view our abundance dataframe, which ranks each ASV by its abundance across samples 


**Question 3C) What are the 5 most abundant ASVs in the dataset?**
**Question 3D) Are these also the most prevalent ones? (Compare your abundance dataframe with your prevalence dataframe)**


- Let's make a Whittaker plot, which plots the abundance of ASVs in the dataset ordered by rank. These can give us an idea of the abundance distribution and how many ASV are dominant vs. rare. 

```{r}
ggplot(merged_df_abundance, aes(x = Rank, y = Abundance)) +
  geom_line() +
  geom_point() +
  labs(
    title = "Rank-Abundance Curve",
    x = "Rank",
    y = "Relative Abundance"
  ) +
  theme_minimal()
```

![image](https://github.com/user-attachments/assets/576683ac-5b90-4262-83ce-ded06f47308c)


**Question 3E) What does this distribution tell you?**

## Part 4 - Explore dataset-specific trends using bar plots

- So far we've explored the metadata and now have a nice overview of what taxa are present in the dataset
- Now we can try to look for trends to gain some insight into the study
- A reminder that this study looks at the effect of diet intervention in mice originating from different vendors
- Lets see if the vendor the mice come from has any effect on composition using the comp_barplot function that we used earlier
- Add the group_by flag to group our plots by a metadata column! (see the function "help" for the command if you can't figure it out)


- **Question 4A) Look through different taxonomic levels - What differences in the relative abundance do you observe based on mouse vendor?**
- **Question 4B) In the published paper they describe differences specifically in Muribaculaceae (polysaccharide-degrading). Describe the differences in this family between vendors**



- It does look like the vendor influences microbiome composition! To determine the effect of diet on these mice, we may have to account for this 
- One strategy is to analyze the effects of diet separately for each vendor. To do this we'll "filter" our phyloseq object for each vendor 
- We can use the subset_samples function to do this
- To make our life easier, subset from the "fixed" ps object 
- For example: To subset for "Beijing" samples...
```{r}
bejing_physeq <- subset_samples(fixed_physeq, vendor == "Beijing")
# view the physeq object to confirm that we're filtering samples as expected!
```


- Now we can start to look at the effect of diet on the microbiomes of these mice... But first there is one more consideration. Time point! 
- 0 = baseline (before diet intervention).  Day 5, 13, 31 are post intervention (Note that experiments performed with certain vendors don't contain all time points)
-  For each vendor, we can investigate the effect of diet at different time points by faceting our plot!
- For simplicity sake, let's just focus on mice from Guangdong and Shanghai at the Family level

```{r}

#For simplicity sake, let's just focus on Guangdong and Shanghai at the Family level

comp_barplot(shanghai_physeq, tax_level = "Family", group_by = "Group", facet_by = "Day", n_taxa = 12)
comp_barplot(guangdong_physeq, tax_level = "Family", group_by = "Group", facet_by = "Day", n_taxa = 12)

```

**Question 4C) In general do you see variability between replicates?**

- **Question 4D) In these studies, cellulose is the control diet (all mice are kept on cellulose before day 0, and then either switch to inulin, resistant starch or are kept on cellulose). Is microbiome composition stable over time in this control? Is this different across taxonomic levels?** 

- **Question 4E) What effects do inulin and resistant starch have on the microbiome? Are these constant over time? Do they differ based on mouse vendor?**



- We can also view the relative abundance of individual taxa to get a more focused look
- This will take some transformation of the data 
- For instance, we can look at the relative abundance of Oscillospiraceeae at time point 31 between the groups in the Shanghai experiment:




```{r}
#Subset our Shanghai phyloseq object for day 31
shanghai_physeq_31 <- subset_samples(shanghai_physeq, Day == 31)


#We can use the tax_transform function to change to relative abundance and get our phyloseq object at the Family level
shanghai_physeq_31_transformed <- tax_transform(shanghai_physeq_31, "compositional", rank = "Family") #


# lets turn this back into a ps object
shanghai_physeq_31_transformed_ps <- ps_get(shanghai_physeq_31_transformed) #returns a phyloseq object 


#now we can add the relative abundance Oscillospiraceae to the sample data column using ps_otu2samdat

shanghai_physeq_31_transformed_ps_osc <- ps_otu2samdat(shanghai_physeq_31_transformed_ps, "Oscillospiraceae") 


#now we can make a table with its abundance 
shanghai_physeq_31_transformed_ps_osc_tab <-   samdat_tbl(shanghai_physeq_31_transformed_ps_osc)



ggplot(shanghai_physeq_31_transformed_ps_osc_tab, aes(x = Group, y = Oscillospiraceae)) +
  geom_boxplot(width = 0.5, colour = "grey35") +
  geom_jitter(width = 0.2, alpha = 0.5) +
  #scale_y_log10() +
  theme_bw()


```


![image](https://github.com/user-attachments/assets/3e75ee79-9356-4570-b50b-c4a32356be82)


**Question 4E) Try to recreate this yourself with a different comparison/taxon that you find interesting and describe how it changes between diet group**



## Part 5 - Data normalization for diversity analyses 

- While our bar plots have given us some nice qualitative measures of the community composition, can we assess changes in a more quantitative way? 

- We can calculate changes in microbial diversity (the number of taxa present and how they are distributed) 

- As covered in the lectures, we can assess within-sample diversity (alpha diversity) and between-sample diversity (beta-diversity). This involves calculating different diversity metrics and calculating ecological distance
-  Before we do this, there is an important consideration: library size normalization! 

- During a sequencing run, different samples will inevitably have different numbers of reads, which means we have to normalize the read depth of our samples to make accurate comparisons


- For the bar plots, we've calculated relative abundance, which essentially "normalized" the samples (reads of taxa/total reads in the sample) to make them comparable. Using relative abundance for normalization in diversity calculations is not generally advised. There is a lot of debate and controversy around what normalization approaches are the best. 

- In reality, no single method is perfect. But we will be using a method called rarefaction. It's relatively simple and a recently published [paper](https://pmc.ncbi.nlm.nih.gov/articles/PMC10900887/) by Schloss put forward that it is the best method for normalizing amplicon sequencing data


- Basically, rarefaction involves the random subsampling of reads in each sample to the SAME depth 

- To illustrate the concept of rarefaction lets look at the distribution of reads in our samples



```{r}
#We can use sample_sums to get the sequencing depth (# of reads per sample)



sequencing_depth <- sample_sums(fixed_physeq)
sequencing_depth_df <- as.data.frame(sequencing_depth)

# ordering the dataframe 
sequencing_depth_ordered  <- sequencing_depth_df[order(sequencing_depth_df$sequencing_depth), ,drop = FALSE ]




```

- **Question 5A) What would be a good sequencing depth to subsample at? (hint: see how you ordered the dataframe in section 3)** 


- Before we subsample, we should ensure that we have enough sequencing depth in our samples. We can do this by plotting rarefaction curves, which tell us if our ability to detect species is saturating or not. 

- Rarefaction curves can be difficult to interpret when we have lots of samples. This is a pretty well sequenced dataset, so low saturation would mostly be an issue only for some of the low sequencing depth. 

- Let's build rarefaction curves for the bottom 20 samples, using the rarecurve function in the vegan package



```{r}
#First let's get a matrix from our phyloseq object
#We are using "t" to transpose the rows and columns (rarecurve takes the taxa and abundance in opposite oritentation to phyloseq)

matrix_phyloseq <- as(t(otu_table(physeq)), "matrix")


## Calculate row sums (sequencing depth of each sample)

row_sums <- rowSums(matrix_phyloseq)

# Order rows by row sums and take the 50 rows with the lowest sums
subset_matrix_phyloseq <- matrix_phyloseq[order(row_sums)[1:20], ]



#now we have a matrix that we can put into the rarecurve function! 
#step = 1000  means that the function calculates ASV richness every 1000 sequences until the total number of species for that sample is reached

rarecurve(subset_matrix_phyloseq, step = 1000, label = FALSE)


```

![image](https://github.com/user-attachments/assets/2579ceff-ab96-4807-8660-8813960b021c)

The curve is built by taking random subsamples every 1,000 reads and calculating the number of ASVs
By default, rarecurve plots "species" on the y-axis (in our case this is ASVs) and sample size refers to the sequencing depth 

Either way it pretty clearly looks looks all samples here are reaching saturation. 



- **Question 5B) In general, what would the lowest acceptable sampling depth where we could have adequate detection of all ASVs?** 



Luckily, our samples are all sequenced quite well. Now that we know that, we can go ahead with our rarefaction. We can subsample so that ALL samples are subsampled to the LOWEST sequencing depth in the dataset. 

This can be done using the rarefy_even_depth function

```{r}
rarefied_phyloseq <- rarefy_even_depth(physeq, sample.size = min(sample_sums(physeq)))

#physloseq object as input, sample.size = the depth to subsample at (in this case the minimum # of reads in samples on our PS object)
```


- Now we have a normalized phyloseq object, where each sample has the same number of reads!
  

## Section 6 - Assessing within-sample diversity 

In this section, we'll explore measurements for determining within-sample diversity (alpha diversity). The two main considerations for ecological diversity are 1) richness (number of ASVs in sample), 2) evenness (how evenly those are distributed). The metrics we use to calculate alpha diversity incorporate either or both of these. 


- We can use phyloseq's plot_richness function. Although it's called plot_richness, it actually can plot a number of different alpha diversity metrics. 

- By using the "observed" flag, we can look at the richness. In this case it will be the number of ASVs.
We can add geom_boxplot to get a nice plot  
- Lets start by plotting richness by Time point (day)



```{r}
plot_richness(rarefied_phyloseq, x="Day", measures=c("Observed")) + 
  geom_boxplot() 
#this looks kinda weird - and we get the warning: Warning: Continuous x aesthetic
# This is probably because the day metadata variable is saved as a continuous variable, let's change this to a "factor" variable
sample_data(rarefied_phyloseq)$Day <- as.factor(sample_data(rarefied_phyloseq)$Day)

plot_richness(rarefied_phyloseq, x="Day", measures=c("Observed")) + 
  geom_boxplot() 

#looks better now!

```

![image](https://github.com/user-attachments/assets/5ba03551-76c1-42ae-9ae0-8b2f6b66c853)



**Question 6A - what is the approximate median number of ASVs per sample at each time point?**


- Now we can move on to diversity metrics that incorporate both richness and evenness. Shannon diversity is more sensitive to changes in richness, while Simpson's is more sensitive to changes in evenness 
- Let's look at the effect of diet Group on these metrics
  

```{r}

plot_richness(rarefied_phyloseq, x="Group", measures=c("Shannon")) + 
  geom_boxplot() 
plot_richness(rarefied_phyloseq, x="Group", measures=c("Simpson")) + 
  geom_boxplot() 
```

**Question 6B) Are there differences between groups based on diet? Are these consistent between diversity metrics? If not, what does this tell you?**


- Because we observed differences based on vendor earlier, this may influence our diet-specific calculations of alpha-diversity

- Lets look at the effects of vendor
   
```{r}
plot_richness(rarefied_phyloseq, x="vendor", measures=c("Shannon")) + 
  geom_boxplot() 
plot_richness(rarefied_phyloseq, x="vendor", measures=c("Simpson")) + 
  geom_boxplot() 
```

- Seems like Shanghai has lower alpha diversity in general! 
- Let's see if the same diet-specific exist for each vendor 
 - Subset your phyloseq objects for each vendor and also  remove the baseline effects so that we can focus simply on the differences between groups post-intervention
 
 
 

```{r}
bejing_physeq_rarefied <- subset_samples(rarefied_phyloseq, vendor == "Beijing" & Day != 0 )
shanghai_physeq_rarefied <- subset_samples(rarefied_phyloseq, vendor == "Shanghai" & Day!= 0)
guangdong_physeq_rarefied <- subset_samples(rarefied_phyloseq, vendor == "Guangdong" & Day!= 0)

#Shannon diversity

plot_richness(bejing_physeq_rarefied, x="Group", measures=c("Shannon")) + 
  geom_boxplot() 
plot_richness(shanghai_physeq_rarefied, x="Group", measures=c("Shannon")) + 
  geom_boxplot() 
plot_richness(guangdong_physeq_rarefied, x="Group", measures=c("Shannon")) + 
  geom_boxplot() 


```


**Question 6C) Are differences in alpha diversity consistent across vendors? across diversity metrics?**




## Part 7 - Between sample diversity 

- One of the best ways to assess changes between microbial communities is by measuring between sample ecological distance 
- There are several distance metrics one can use to assess differences. Regardless, these metrics generally work by measuring the richness and/or the eveness of the community. 
- While others are used, generally there are 4 main distance metrics used in microbial ecology analyses: 1) Bray-Curtis (richness + eveness), 2) Weighted Unifrac distance (richness + eveness + phylogenetic relatedness), 3) Jaccard distance (richness), 4) Unweighted Unifrac distance (richness + phylogenetic relatedness).


- Lets start by calculating distance using the distance function in phyloseq
- You can list all available distance metrics that this function is compatible using:

```{r}
distanceMethodList
```
- Note: Weighted unifrac = wunifrac, unweighted = unifrac

- To calculate bray-curtis distance on our rarefied phyloseq object we can...
  

```{r}
bray_distance <- distance(rarefied_phyloseq, method = "bray")

```

- This returns a matrix with the distances between samples shown. For Bray-Curtis distance, all values will be between 0 and 1

- There are plenty of things that one can do with a distance matrix, but the most common is to perform an ordination. This process will essentially take our distance matrix and represent our samples in a two-dimensional space, where samples with less distance between each other are closer together and samples with greater distances are further apart. 

- There are lots of different ordination methods, but we'll mostly focus on principal coordinate analysis (PCoA) plots. These 1) take your distance matrix, 2) Performs a "decomposition", which projects the data into 2-D space while trying to best represent the distances between samples. The mathematical details are not important for our purposes, but the important thing to know is that we'll end up with two axes that represent variance in the data. The variance explained is proportional to the distances between samples. 

- This will make sense if we perform an ordination! 

- We can plot can calculate the distance and perform an ordination in two steps:

```{r}
#create ordination 
# takes our method of phyloseq object, ordination method, and 

ordination_pcoa_bray <- ordinate(rarefied_phyloseq, "PCoA", "bray")

#now we can plot this 
plot_ordination(rarefied_phyloseq, ordination_pcoa_bray)

```

![image](https://github.com/user-attachments/assets/c861faf1-2315-4ec1-be20-5af78ae07062)




- We have a PCoA plot now! 
**Question 7A) Imagine two points on the PCoA plot above. In case one, these two points are distributed vertically and in case 2, the points are distributed horizontally (equidistant to case 1). In which case is there greater distance between points?**

- Since each sample represents a different point, we see how our samples cluster based on experimental metadata by assigning colours to different variables


- generate plots between the different metdata groups by replacing "metadata group" with the different metadata groups. 

```{r}
plot_ordination(rarefied_phyloseq, ordination_pcoa_bray, color = "" )

```

**Question 7B) Which metadata has the largest effect on beta-diversity based on Bray-Curtis ordinations?**


- Let's focus on vendor-specific differences. Recreate the vendor-coloured PCoA, but this time using weighted unifrac distance instead of Bray-Curtis distance


**Question 7C) Qualitatively, which distance metric yields higher differences between the groups. Why would this be the case?**


- Like we did in the previous sections, let's look at the effect of diet on each vendor's microbiome. Again, subset your phyloseq object for each vendor and generate Bray-Curtis PCoAs and remove the baseline effects so that we can focus simply on the differences between groups post-intervention

```{r}

#subset for each vendor 
# can use the != operator to select samples that ARE NOT day 0 (baseline)
bejing_physeq_rarefied <- subset_samples(rarefied_phyloseq, vendor == "Beijing" & Day != 0 )
shanghai_physeq_rarefied <- subset_samples(rarefied_phyloseq, vendor == "Shanghai" & Day!= 0)
guangdong_physeq_rarefied <- subset_samples(rarefied_phyloseq, vendor == "Guangdong" & Day!= 0)


#create ordinations for each vendor 
ordination_pcoa_bray_beijing <- ordinate(bejing_physeq_rarefied, "PCoA", "bray")
ordination_pcoa_bray_shanghai <- ordinate(shanghai_physeq_rarefied, "PCoA", "bray")
ordination_pcoa_bray_guangdong <- ordinate(guangdong_physeq_rarefied, "PCoA", "bray")



#plot 
plot_ordination(bejing_physeq_rarefied, ordination_pcoa_bray_beijing, color = "Group")
plot_ordination(shanghai_physeq_rarefied, ordination_pcoa_bray_shanghai, color = "Group")
plot_ordination(guangdong_physeq_rarefied, ordination_pcoa_bray_guangdong, color = "Group")

```

**Question 7D) In which vendor is the diet effects most pronounced? Does resistant start or inulin generally have a larger effect compared to the control cellulose diet?**

**Question 7E) In the shanghai vendor, try to determine if the effects of inulin are static, or change over time (hint: you'll need to play around with filtering)**

- So far, we've relied on our eyes to test for differences. Can we do this more quantiatively? The PERMANOVA test is multivariate test commonly used to test for differences between distance matrices. Significance in a PERMANOVA test implies that the distances WITHIN a group are significantly LESS than the differences BETWEEN at least one group.

- We'll need to perform this test on a distance matrix. Let's start by performing this on the whole dataset 


```{r}
# Calculate bray curtis distance matrix
bray_distance <- distance(rarefied_phyloseq, method = "bray")

# make a data frame from the sample_data sow we can feed it into 
metadata_df <- data.frame(sample_data(rarefied_phyloseq))

# Adonis PERMANOVA test
#Requires,the bray_distance matrix, the group of interest to compare and the metadata dataframe 

adonis(bray_distance ~ Group, data = metadata_df)
```

![image](https://github.com/user-attachments/assets/29237dcf-102c-4f16-912f-ccdf2a15ea3b)


Here, we can see the output of the adonis test. The two important columns for us are the p-value column, which tells us that diet group does have a significant effect! The R2 column tells us what proportion of total variance of the data is explained by a certain factor (diet group) in this case. 


**Question 7F) Which metadata groups have a significant effect on beta-diversity? Which has the largest effect?**



- Similar to our PCoA plots, lets look at the effects of diet within each vendor group 

```{r}
bejing_physeq_rarefied <- subset_samples(fixed_physeq, vendor == "Beijing" & Day != 0 )
shanghai_physeq_rarefied <- subset_samples(fixed_physeq, vendor == "Shanghai" & Day!= 0)
guangdong_physeq_rarefied <- subset_samples(fixed_physeq, vendor == "Guangdong" & Day!= 0)


# Calculate bray curtis distance matrix
bray_distance_beijing <- distance(bejing_physeq_rarefied, method = "bray")
bray_distance_shanghai <- distance(shanghai_physeq_rarefied, method = "bray")
bray_distance_guangdong <- distance(guangdong_physeq_rarefied, method = "bray")

# make a data frame from the sample_data sow we can feed it into 
metadata_df_b <- data.frame(sample_data(bejing_physeq_rarefied))
metadata_df_s <- data.frame(sample_data(shanghai_physeq_rarefied))
metadata_df_g <- data.frame(sample_data(guangdong_physeq_rarefied))


adonis(bray_distance_beijing ~ Group, data = metadata_df_b)
adonis(bray_distance_shanghai ~ Group, data = metadata_df_s)
adonis(bray_distance_guangdong ~ Group, data = metadata_df_g)



```


**Question 7E) In which vendor is the effect of diet the largest? Does this match your observations with the corresponding PCoA plot?**



***Optional section - using PCAs to determine important taxa
- Since PCoAs work on distance matrices, they don't have any information on taxonomy. However, principal component analysis (PCA) are formed using transformed (centered-log-transformed) microbial counts. This way, we can see which taxa are contributing to differences in the community. 
- We can plot different genera as vectors, where the length of the vector is proportional to its contribution to the PCA axis
- Lets do this for the whole community, coloured by vendor 


```{r}

rarefied_phyloseq %>%
  tax_fix(unknowns = c("uncultured")) %>%
  tax_transform("clr", rank = "Genus") %>%
  # when no distance matrix or constraints are supplied, PCA is the default/auto ordination method
  ord_calc(method = "PCA") %>%
  ord_plot(color = "vendor", plot_taxa = 1:5, size = 2) +
  scale_colour_brewer(palette = "Dark2")



```

![image](https://github.com/user-attachments/assets/bfdd99e2-fb3d-4ab8-8cde-001fad2dcff5)


- We can for instance that Tannerellaceae is contributing to the differences in the Shanghai-specific community
- We can play around with number of taxa shown using the plot_taxa flag

***Optional Question 7F) What genera are driving diet differences for each vendor**


## Summary 

- That's all for the walkthrough. Again, this is just the tip of the iceberg. There are plenty additional statistical approaches that can be applied. But, hopefully this is has been helpful in getting a broad overview of what is possible when analyzing 16S data

**Question 8) Now that we've had the chance to analyze the data in a number of different ways, using a few of the methods we've covered in the walkthrough, try to describe how diet effects microbiome composition in this study**


